### PR TITLE
extmod/modbluetooth: Make UUID support the buffer protocol.

### PR DESCRIPTION
**Pull request by jimmo**
This allows UUIDs to be easily used in BLE payloads (such as advertising).

Add this to the advertising example.

Ref: #5186

### CODE REVIEW

File: extmod/modbluetooth.c [hunk 1]
```diff
@@ -172,6 +172,37 @@ STATIC void bluetooth_uuid_print(const mp_print_t *print, mp_obj_t self_in, mp_p
...
+        bufinfo->buf = buf;
+        bufinfo->len = self->type;
+        bufinfo->typecode = 'B';
+        return 0;
+    } else {
+        // can't write to a string
+        bufinfo->buf = NULL;
+        bufinfo->len = 0;
+        bufinfo->typecode = -1;
+        return 1;
```
Comment (dpgeorge) [inline extmod/modbluetooth.c]: It should be enough to just return 1 here, not bother to set any fields.

(I see that you may have followed what's done in py/objstr.c... so that code there could probably be simplified as well to just return 1.)

File: extmod/modbluetooth.c [hunk 2]
```diff
@@ -172,6 +172,37 @@ STATIC void bluetooth_uuid_print(const mp_print_t *print, mp_obj_t self_in, mp_p
     }
 }
 
+mp_int_t bluetooth_uuid_get_buffer(mp_obj_t self_in, mp_buffer_info_t *bufinfo, mp_uint_t flags) {
+    mp_obj_bluetooth_uuid_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (flags == MP_BUFFER_READ) {
+        uint8_t *buf = m_new(uint8_t, self->type);
```
Comment (dpgeorge) [inline extmod/modbluetooth.c]: get_buffer usually (I think always?) returns just a reference into the data, not a copy of it.  For efficiency reasons, and the caller always knows that it's just a reference (and for writable buffers allows the caller to modify it inplace).

So, for the 128-bit UUID it's simple to just do `bufinfo->buf = self->uuid._128`.

For 16-bit and 32-bit it might be possible to just return the address of the `_16` or `_32` entry.. but that probably won't be portable across archs with different endianness.

But looking at the code, it might be possible to simplify the UUID class a lot by always making it store the UUID value in a `uint8_t` array.  For 16 and 32 bit values it'll sometimes need to pack/unpack the value, but I think it'll simplify a lot of code paths that use UUID's, especially the one here.

File: examples/bluetooth/ble_advertising.py [hunk 3]
```diff
@@ -11,8 +11,14 @@
 _ADV_TYPE_FLAGS = const(0x01)
 _ADV_TYPE_NAME = const(0x09)
 _ADV_TYPE_UUID16_COMPLETE = const(0x3)
+_ADV_TYPE_UUID32_COMPLETE = const(0x5)
+_ADV_TYPE_UUID128_COMPLETE = const(0x7)
+_ADV_TYPE_UUID16_MORE = const(0x2)
+_ADV_TYPE_UUID32_MORE = const(0x4)
+_ADV_TYPE_UUID128_MORE = const(0x6)
```
Comment (dpgeorge) [inline examples/bluetooth/ble_advertising.py]: These updates to the examples should be in separate commits (I guess they got added by accident here?).

**Comment by jimmo** [comment]
Thanks, addressed all the comments and split the commits.

The UUID is now internally LE, the public interface is still the same.

Added the objstr change as another commit.

File: extmod/modbluetooth.c [hunk 4]
```diff
@@ -155,13 +148,13 @@ STATIC void bluetooth_uuid_print(const mp_print_t *print, mp_obj_t self_in, mp_p
...
-        mp_printf(print, "UUID16(0x%04x)", self->uuid._16);
+        mp_printf(print, "UUID16(0x%02x%02x)", self->data[1], self->data[0]);
     } else if (self->type == MP_BLUETOOTH_UUID_TYPE_32) {
-        mp_printf(print, "UUID32(0x%08x)", self->uuid._32);
+        mp_printf(print, "UUID32(0x%02x%02x%02x%02x)", self->data[3], self->data[2], self->data[1], self->data[0]);
     } else if (self->type == MP_BLUETOOTH_UUID_TYPE_128) {
         mp_printf(print, "UUID128('");
         for (int i = 0; i < 16; ++i) {
-            mp_printf(print, "%02x", self->uuid._128[15-i]);
+            mp_printf(print, "%02x", self->data[15-i]);
```
Comment (dpgeorge) [inline extmod/modbluetooth.c]: Could the printing be combined into one, with the first bit like:

    mp_printf(print, "UUID%u(%s", self->type * 8, self->type <=4 ? "0x" : "'");

File: extmod/modbluetooth_nimble.c [hunk 5]
```diff
@@ -115,23 +116,27 @@ STATIC mp_obj_bluetooth_uuid_t create_mp_uuid(const ble_uuid_any_t *uuid) {
...
         case BLE_UUID_TYPE_16:
             result.type = MP_BLUETOOTH_UUID_TYPE_16;
-            result.uuid._16 = uuid->u16.value;
+            result.data[0] = uuid->u16.value & 0xff;
+            result.data[1] = (uuid->u16.value << 8) & 0xff;
             break;
         case BLE_UUID_TYPE_32:
             result.type = MP_BLUETOOTH_UUID_TYPE_32;
-            result.uuid._32 = uuid->u32.value;
+            result.data[0] = uuid->u16.value & 0xff;
```
Comment (dpgeorge) [inline extmod/modbluetooth_nimble.c]: u16 should be u32

File: py/objstr.c [hunk 6]
```diff
@@ -1912,10 +1912,7 @@ mp_int_t mp_obj_str_get_buffer(mp_obj_t self_in, mp_buffer_info_t *bufinfo, mp_u
         bufinfo->typecode = 'B'; // bytes should be unsigned, so should unicode byte-access
         return 0;
     } else {
-        // can't write to a string
-        bufinfo->buf = NULL;
-        bufinfo->len = 0;
-        bufinfo->typecode = -1;
+        // can't write to a string.
```
Comment (dpgeorge) [inline py/objstr.c]: Minor thing: please don't update comments if they don't need it, like this.  (Yes comments are a bit inconsistent in their formatting, but that's a separate issue :)

**Comment by jimmo** [comment]
Thanks, done!

File: extmod/modbluetooth.c [hunk 7]
```diff
@@ -135,83 +136,57 @@ STATIC mp_obj_t bluetooth_uuid_unary_op(mp_unary_op_t op, mp_obj_t self_in) {
...
-        mp_printf(print, "UUID128('");
-        for (int i = 0; i < 16; ++i) {
-            mp_printf(print, "%02x", self->uuid._128[15-i]);
-            if (i == 3 || i == 5 || i == 7 || i == 9) {
-                mp_printf(print, "-");
-            }
+    mp_printf(print, "UUID%u(%s", self->type * 8, self->type <= 4 ? "0x" : "'");
+    for (int i = 0; i < self->type; ++i) {
+        mp_printf(print, "%02x", self->data[self->type - 1 - i]);
+        if (self->type == 16 && (i == 3 || i == 5 || i == 7 || i == 9)) {
```
Comment (dpgeorge) [inline extmod/modbluetooth.c]: This could go one line up, the `self->type==16` check removed, and 3/5/7/9 all increased by one.

File: extmod/modbluetooth.c [hunk 8]
```diff
@@ -135,83 +136,57 @@ STATIC mp_obj_t bluetooth_uuid_unary_op(mp_unary_op_t op, mp_obj_t self_in) {
...
+            mp_printf(print, "-");
         }
-        mp_printf(print, "')");
-    } else {
-        mp_printf(print, "UUID?(%d)", self->type);
     }
+    if (self->type == 16) {
```
Comment (dpgeorge) [inline extmod/modbluetooth.c]: Probable good idea to use `MP_BLUETOOTH_UUID_TYPE_128`...?

**Comment by jimmo** [comment]
Done

**Comment by dpgeorge** [comment]
Thanks!
